Could not compress the PHAR: the compression requires too many file descriptors to be opened

## Bug report

| Question    | Answer
| ------------| ---------------
| Box version |3.8.0
| PHP version | 7.2.18
| Platform with version | OSX

When trying to compile Box, I got:

```
 [ERROR] Could not compress the PHAR: the compression requires too many file descriptors to be opened (1806). Check your
         system limits or install the posix extension to allow Box to automatically configure it during the compression
```

This is the first time ever I got it... 1806 matches the number of files found in `box.phar`, but what looks weird is that:

```
posix_getrlimit()['soft openfiles'] => 10240
posix_getrlimit()['hard openfiles'] => 10240
```

So it is likely that either checking this file limit is not enough...

For the record, the limit is bumped [here](https://github.com/humbug/box/blob/master/src/Console/Command/Compile.php#L621) and the actual code for bumping the limit is [here](https://github.com/humbug/box/blob/master/src/functions.php#L277)

--- a/src/Console/Command/Compile.php
+++ b/src/Console/Command/Compile.php
@@ -190,7 +190,16 @@ HELP;
 
         $this->removeExistingArtifacts($config, $logger, $debug);
 
-        $box = $this->createPhar($config, $logger, $io, $debug);
+        // Adding files might result in opening a lot of files. Either because not parallelized or when creating the
+        // workers for parallelization.
+        // As a result, we bump the file descriptor to an arbitrary number to ensure this process can run correctly
+        $restoreLimit = bump_open_file_descriptor_limit(2048, $io);
+
+        try {
+            $box = $this->createPhar($config, $logger, $io, $debug);
+        } finally {
+            $restoreLimit();
+        }
 
         $this->correctPermissions($path, $config, $logger);
 
@@ -618,7 +627,7 @@ HELP;
             )
         );
 
-        $restoreLimit = bump_open_file_descriptor_limit($box, $io);
+        $restoreLimit = bump_open_file_descriptor_limit(count($box), $io);
 
         try {
             $extension = $box->compress($algorithm);

--- a/src/Console/Command/Extract.php
+++ b/src/Console/Command/Extract.php
@@ -14,6 +14,7 @@ declare(strict_types=1);
 
 namespace KevinGH\Box\Console\Command;
 
+use function count;
 use KevinGH\Box\Box;
 use function KevinGH\Box\bump_open_file_descriptor_limit;
 use KevinGH\Box\Console\IO\IO;
@@ -97,7 +98,7 @@ final class Extract extends BaseCommand
             return 1;
         }
 
-        $restoreLimit = bump_open_file_descriptor_limit($box, $io);
+        $restoreLimit = bump_open_file_descriptor_limit(count($box), $io);
 
         $outputDir = $input->getArgument(self::OUTPUT_ARG);
 

--- a/src/functions.php
+++ b/src/functions.php
@@ -274,9 +274,9 @@ function register_error_handler(): void
  *
  * @return Closure callable to call to restore the original maximum number of open files descriptors
  */
-function bump_open_file_descriptor_limit(Box $box, IO $io): Closure
+function bump_open_file_descriptor_limit(int $count, IO $io): Closure
 {
-    $filesCount = count($box) + 128;  // Add a little extra for good measure
+    $count += 128;  // Add a little extra for good measure
 
     if (false === function_exists('posix_getrlimit') || false === function_exists('posix_setrlimit')) {
         $io->writeln(
@@ -291,7 +291,7 @@ function bump_open_file_descriptor_limit(Box $box, IO $io): Closure
     $softLimit = posix_getrlimit()['soft openfiles'];
     $hardLimit = posix_getrlimit()['hard openfiles'];
 
-    if ($softLimit >= $filesCount) {
+    if ($softLimit >= $count) {
         return static function (): void {};
     }
 
@@ -301,7 +301,7 @@ function bump_open_file_descriptor_limit(Box $box, IO $io): Closure
             .'</info>',
             $softLimit,
             $hardLimit,
-            $filesCount,
+            $count,
             'unlimited'
         ),
         OutputInterface::VERBOSITY_DEBUG
@@ -309,7 +309,7 @@ function bump_open_file_descriptor_limit(Box $box, IO $io): Closure
 
     posix_setrlimit(
         POSIX_RLIMIT_NOFILE,
-        $filesCount,
+        $count,
         'unlimited' === $hardLimit ? POSIX_RLIMIT_INFINITY : $hardLimit
     );
 

